File: AI_UoB/programs/data_structures_for_ai/4_queue.c
#include <stdio.h>
#include <stdlib.h>

#define MAX 100

int queue[MAX];
int front = -1;
int rear = -1;

void enqueue(int value) {
    if (rear == MAX - 1) {
        printf("Queue is Full\n");
        return;
    }
    if (front == -1)
        front = 0;
    queue[++rear] = value;
}

void dequeue() {
    if (front == -1) {
        printf("Queue is Empty\n");
        return;
    }
    printf("Dequeued: %d\n", queue[front++]);
    if (front > rear)
        front = rear = -1;
}

int peek() {
    if (front == -1) {
        printf("Queue is Empty\n");
        return -1;
    }
    return queue[front];
}

int isEmpty() {
    return front == -1;
}

void display() {
    if (front == -1) {
        printf("Queue is Empty\n");
        return;
    }
    printf("FRONT -> ");
    for (int i = front; i <= rear; i++)
        printf("%d -> ", queue[i]);
    printf("REAR\n");
}

int main() {
    enqueue(10);
    enqueue(20);
    enqueue(30);
    enqueue(40);
    display();
    printf("Peek: %d\n", peek());
    dequeue();
    display();
    printf("Empty: %d\n", isEmpty());
    return 0;
}

File: AI_UoB/programs/data_structures_for_ai/4_queue.py
class Queue:
    def __init__(self):
        self.queue = []

    def enqueue(self, value):
        self.queue.append(value)

    def dequeue(self):
        if self.isEmpty():
            print("Queue is Empty")
            return
        print(f"Dequeued: {self.queue.pop(0)}")

    def peek(self):
        if self.isEmpty():
            print("Queue is Empty")
            return
        return self.queue[0]

    def isEmpty(self):
        return len(self.queue) == 0

    def display(self):
        if self.isEmpty():
            print("Queue is Empty")
            return
        print("FRONT -> " + " -> ".join(str(x) for x in self.queue) + " -> REAR")

if __name__ == "__main__":
    q = Queue()
    q.enqueue(10)
    q.enqueue(20)
    q.enqueue(30)
    q.enqueue(40)
    q.display()
    print(f"Peek: {q.peek()}")
    q.dequeue()
    q.display()
    print(f"Empty: {q.isEmpty()}")